src/Microsoft.Tye.Hosting/ReplicaStateRecorder.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Tye.Hosting.Model;
+
+namespace Microsoft.Tye.Hosting
+{
+    public class ReplicaStateRecorder : IApplicationProcessor
+    {
+        private readonly ILogger _logger;
+        private ReplicaRegistry _registry;
+
+        public ReplicaStateRecorder(Model.Application application, ILogger logger, IDictionary<ServiceType, IReplicaInstantiator> replicaInstantiators)
+        {
+            _logger = logger;
+            _registry = new ReplicaRegistry(application, replicaInstantiators);
+        }
+
+        public async Task StartAsync(Model.Application application)
+        {
+            await _registry.Reset();
+
+            foreach (var service in application.Services.Values)
+            {
+                service.Items[typeof(Subscription)] = service.ReplicaEvents.Subscribe(OnReplicaChanged);
+            }
+        }
+
+        public Task StopAsync(Model.Application application)
+        {
+            foreach (var service in application.Services.Values)
+            {
+                if (service.Items.TryGetValue(typeof(Subscription), out var item) && item is IDisposable disposable)
+                {
+                    disposable.Dispose();
+                }
+            }
+
+            _registry.Dispose();
+            return Task.CompletedTask;
+        }
+
+        private void OnReplicaChanged(ReplicaEvent replicaEvent)
+        {
+            OnReplicaChangedAsync(replicaEvent).Wait(); // :(
```

Comment:
`:( ` is correct 😆 

It's probably a better fix to make ` _registry.WriteReplicaEvent` synchronous. The disavantage of `Task.Wait` is that it uses two threads (the thread that waits, and the thread to does the underlying async operation). Just using `File.WriteAllText` for this case is going to be better.

The other option would be to fire-and-forget, but just being synchronous sounds better.